### PR TITLE
Polars: error if string given as margin identifier

### LINKED ISSUE #1977: Usability: When grouping by single columns, easy to forget the closing comma in the tuple
It's easy to type
```
("ILOSTAT"): dp.polars.Margin(public_info="keys", max_partition_length=60_000_000)
```
when you mean
```
("ILOSTAT", ): dp.polars.Margin(public_info="keys", max_partition_length=60_000_000)
```
(note the trailing comma that makes it a tuple.)

Downsteam, it complains that you didn't specify a margin, but it really looks like you did!

Options:
- Silently transform strings to tuples?
- Warn earlier if there is a non-tuple key?

### PR DESCRIPTION
**Pull request by Shoeboxam**
Close #1977

### CODE REVIEW

**Comment by Shoeboxam** [comment]
* **#2089** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2089?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python/src/opendp/context.py [hunk 1]
```diff
@@ -403,6 +403,8 @@ def compositor(
 
         if margins:
             for by, margin in margins.items():
+                # wrap single strings in a singleton tuple
+                by = (by,) if isinstance(by, str) else by
```
Comment (mccalluc) [inline python/src/opendp/context.py:407]: Possibly steer users towards the tuple syntax?
```suggestion
                if isinstance(by, str):
                    from warnings import warn
                    warn("Tuples preferred for margin keys. For single-valued tuples, include a trailing comma, ie: ('my-key',)")
                    by = (by,)
```

**Review by mccalluc — DISMISSED** [review]
Approving, though I would have leaned towards adding an earlier validation instead. Adding a warning is just an idea: Either way is fine.

File: python/src/opendp/context.py (hunk 1 above)
Comment (Shoeboxam) [reply python/src/opendp/context.py]: I changed the PR to raise an exception if the input is not a tuple, and to use the grouping key in the error message if it is a string.

File: python/src/opendp/context.py [hunk 2]
```diff
@@ -403,6 +403,10 @@ def compositor(
 
         if margins:
             for by, margin in margins.items():
+                # wrap single strings in a singleton tuple
+                if not isinstance(by, tuple):
+                    msg = by if isinstance(by, str) else "your-column"
+                    raise ValueError(f"Margin keys must be tuples. For single-valued tuples include a trailing comma, ie: `('{msg}',)`")
```
Comment (mccalluc) [inline python/src/opendp/context.py:409]: Double check that our coverage checks are working correctly? The ternary introduces a new branch, and I'm a little surprised that that isn't being flagged... but maybe it doesn't count branches within lines?

**Review by mccalluc — COMMENTED** [review]
I'm happy with this direction. Retitle the PR, since you decided not to go that way?

File: python/src/opendp/context.py [hunk 3]
```diff
@@ -382,7 +382,7 @@ def compositor(
         split_evenly_over: Optional[int] = None,
         split_by_weights: Optional[list[float]] = None,
         domain: Optional[Domain] = None,
-        margins: Optional[dict[tuple[str, ...], Margin]] = None,
+        margins: Optional[dict[Union[tuple[str, ...], str], Margin]] = None,
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply python/src/opendp/context.py:385]: I love your attention to detail.

**Review by mccalluc — COMMENTED** [review]
Last commit looks good.

Filed a follow-up issue for a fix to the typing across the codebase.
- #2164

**Review by mccalluc — APPROVED** [review]
Sorry, meant to approve.

**Pull request merged**